CentaurusInfra/mizar
# [Fornax-Mizar-Integration] Added gateway and virtual to subnet in data plane

There are following changes
1. Since we added virtual and cluster gateway to control plane in https://github.com/CentaurusInfra/mizar/pull/582, we made the data plane changes as well in trn cli to get virtual and cluster values. Tests are added as well.

2. There are changes in transit xdp:
a. Don't update cache if the requests come from virtual subnets. The request might come from remote clusters. It is not needed to update cache to set up direct path.

b. If the request targets are cluster gateway, pass the request to user space gateway agents. Since all the virtual subnets do not have endpoints in the the cluster, all the requests aiming virtual subnets have to pass to a new gateway agents to process.

The change will not affect the current mizar logic if the requests are not virtual subnet related.
Detailed tests have to wait when gateway agent is merged. 

Manual test steps:
1. Create a cluster with 2 worker nodes and 1 cluster gateway.
2. Create a vpc vpc1 and one local subnet  net1    and one virtual subnet net2
```bash
root@ip-172-31-7-100:~/mizar_cluster_scripts# kubectl get subnets
NAME   IP              PREFIX   VNI       VPC    STATUS        VIRTUAL   BOUNCERS   CREATETIME                   PROVISIONDELAY
net0   20.0.0.0        8        1         vpc0   Provisioned   false     1          2022-01-12T21:53:13.216790   81.060556
net1   192.168.0.0     24       9044412   vpc1   Provisioned   false     1
net2   192.168.122.0   24       9044412   vpc1   Provisioned   true      1
```
3. Create a pod in subnet net1
```bash
root@ip-172-31-7-100:~/mizar_cluster_scripts# kubectl get pods pod-in-net1 -o wide
NAME          READY   STATUS    RESTARTS   AGE     IP            NODE              NOMINATED NODE   READINESS GATES
pod-in-net1   1/1     Running   0          6m34s   192.168.0.6   ip-172-31-6-231   <none>           <none>
```
4. Ping a virtual subnet ip 
```bash
root@ip-172-31-7-100:~/mizar_cluster_scripts# ./blind_ping_122.sh
-------
pinging 192.168.122.13 from 192.168.0.6
-------
PING 192.168.122.13 (192.168.122.13) 56(84) bytes of data.
```
5. Verify the request targeting gateway host  is redirected to user space
```bash
cat /sys/kernel/debug/tracing/trace_pipe | tee out
```
According to the code changes in src/xdp/trn_transit_xdp.c, if the request is redirect to user spaces, it will print out "bpf_debug("[Transit:%d:0x%x] Edge gateway\n", __LINE__, bpf_ntohl(pkt->inner_ip->daddr));" as below

```bash
          <idle>-0     [000] ..s. 85815.136730: 0: [Transit:475:0xc0a87a0d] Edge gateway
          <idle>-0     [000] ..s. 85816.142070: 0: [Transit:751:0xac1f02d9] Processing IP
          <idle>-0     [000] ..s. 85816.142106: 0: [Transit:475:0xc0a87a0d] Edge gateway
          <idle>-0     [000] ..s. 85817.165908: 0: [Transit:751:0xac1f02d9] Processing IP
          <idle>-0     [000] ..s. 85817.165943: 0: [Transit:475:0xc0a87a0d] Edge gateway
```
